File: apex_dashboard/apex_dashboard/page/liabilities_dashboard/liabilities_dashboard.py
from __future__ import annotations

from typing import Dict, List, Optional, Sequence, Tuple

import frappe
from frappe import _
from frappe.utils import getdate, nowdate

from apex_dashboard.dashboard import utils as dashboard_utils


@frappe.whitelist()
def get_context(context):
	context.no_cache = 1
	context.page_title = _("Liabilities Dashboard")
	return context


@frappe.whitelist()
def get_dashboard_data(company: Optional[str] = None, posting_date: Optional[str] = None):
	try:
		posting_date = posting_date or nowdate()
		posting_date = str(getdate(posting_date))

		data = build_liabilities_snapshot(company=company, posting_date=posting_date)

		return {
			"success": True,
			"filters": {
				"company": company,
				"posting_date": posting_date,
			},
			"data": data,
		}
	except Exception as exc:
		frappe.log_error(frappe.get_traceback(), "Liabilities Dashboard - get_dashboard_data")
		return {
			"success": False,
			"error": _(str(exc)),
		}


def build_liabilities_snapshot(company: Optional[str], posting_date: str) -> Dict[str, object]:
	mapping = dashboard_utils.get_accounts_for_section("liabilities_dashboard")

	sections: List[Dict[str, object]] = []

	suppliers_section = _build_section(
		title=_("الموردون وشركات الخدمات"),
		entries=[
			(_("الموردون"), mapping.get("suppliers")),
			(_("شركات الشحن"), mapping.get("shipping")),
			(_("الجمارك"), mapping.get("customs")),
		],
		company=company,
		posting_date=posting_date,
	)
	sections.append(suppliers_section)

	employees_section = _build_section(
		title=_("مستحقات الموظفين والشركاء"),
		entries=[
			(_("رواتب الموظفين"), mapping.get("employees")),
			(_("شركاء المبيعات"), mapping.get("partners")),
			(_("دائنون آخرون"), mapping.get("other_creditors")),
		],
		company=company,
		posting_date=posting_date,
	)
	sections.append(employees_section)

	notes_section = _build_section(
		title=_("Notes Payable"),
		entries=[(_("Notes Payable"), mapping.get("notes_payable"))],
		company=company,
		posting_date=posting_date,
	)
	sections.append(notes_section)

	taxes_section = _build_section(
		title=_("الضرائب والجهات السيادية"),
		entries=[(_("الضرائب والرسوم"), mapping.get("taxes"))],
		company=company,
		posting_date=posting_date,
	)
	sections.append(taxes_section)

	other_section = _build_section(
		title=_("التزامات أخرى"),
		entries=[
			(_("التزامات المخزون"), mapping.get("stock_liabilities")),
			(_("أرصدة دائنة متنوعة"), mapping.get("credit_balances")),
			(_("السحب على المكشوف"), mapping.get("overdrafts")),
		],
		company=company,
		posting_date=posting_date,
	)
	sections.append(other_section)

	kpis = _build_kpis(suppliers_section, taxes_section, employees_section, notes_section, other_section)
	alerts = _generate_liability_alerts(taxes_section, employees_section, other_section)

	return {
		"kpis": kpis,
		"sections": sections,
		"alerts": alerts,
	}


def _build_section(
	title: str,
	entries: Sequence[Tuple[str, object]],
	company: Optional[str],
	posting_date: str,
) -> Dict[str, object]:
	groups: List[Dict[str, object]] = []
	total_currency: Dict[str, float] = {}
	total_base = 0.0

	for label, config in entries:
		accounts = _normalize_accounts(config)
		balances = dashboard_utils.get_balances_for_accounts(accounts, company=company, posting_date=posting_date)
		by_currency = dashboard_utils.summarize_balances_by_currency(balances)
		base_total = dashboard_utils.summarize_base_total(balances)

		for currency, amount in by_currency.items():
			total_currency[currency] = total_currency.get(currency, 0.0) + amount
		total_base += base_total

		groups.append(
			{
				"label": label,
				"accounts": accounts,
				"balances": [balance.__dict__ for balance in balances],
				"totals": {
					"by_currency": by_currency,
					"base": base_total,
				},
			}
		)

	return {
		"title": title,
		"groups": groups,
		"totals": {
			"by_currency": total_currency,
			"base": total_base,
		},
	}


def _normalize_accounts(config: object) -> List[str]:
	if isinstance(config, dict) and "accounts" in config:
		return config.get("accounts") or []
	if isinstance(config, dict):
		return []
	if isinstance(config, (list, tuple)):
		return list(config)
	if config:
		return [config]
	return []


def _build_kpis(*sections: Dict[str, object]) -> List[Dict[str, object]]:
	kpis: List[Dict[str, object]] = []
	names = ["suppliers", "taxes", "employees", "notes", "others"]
	labels = [
		_("إجمالي الموردين"),
		_("إجمالي الضرائب"),
		_("مستحقات الموظفين"),
		_("Notes Payable"),
		_("التزامات أخرى"),
	]

	for idx, section in enumerate(sections):
		if not section:
			continue
		totals = section.get("totals") or {}
		kpis.append(
			{
				"key": names[idx],
				"label": labels[idx],
				"totals": totals,
				"indicator": "danger" if (totals.get("base") or 0) > 0 else "info",
			}
		)

	return kpis


def _generate_liability_alerts(
	taxes_section: Dict[str, object],
	employees_section: Dict[str, object],
	other_section: Dict[str, object],
) -> List[Dict[str, object]]:
	alerts: List[Dict[str, object]] = []

	tax_total = (taxes_section.get("totals") or {}).get("base", 0) if taxes_section else 0
	if tax_total and tax_total > 0:
		alerts.append(
			{
				"level": "warning",
				"message": _("هناك ضرائب مستحقة الدفع، راجع جدول السداد."),
			}
		)

	employee_total = (employees_section.get("totals") or {}).get("base", 0) if employees_section else 0
	if employee_total and employee_total > 0:
		alerts.append(
			{
				"level": "danger",
				"message": _("لا تزال مستحقات الموظفين معلقة، يُرجى صرفها قبل موعد الرواتب."),
			}
		)

	overdraft_total = 0
	if other_section:
		for group in other_section.get("groups", []):
			if _("السحب على المكشوف") in group.get("label", ""):
				overdraft_total = group.get("totals", {}).get("base", 0)
				break

	if overdraft_total and overdraft_total > 0:
		alerts.append(
			{
				"level": "warning",
				"message": _("هناك استخدام للسحب على المكشوف، راجع خطط التمويل."),
			}
		)

	if not alerts:
		alerts.append(
			{
				"level": "info",
				"message": _("لا توجد تنبيهات حرجة في الوقت الحالي."),
			}
		)

	return alerts

File: apex_dashboard/apex_dashboard/page/liabilities_dashboard/liabilities_dashboard.js
frappe.pages["liabilities-dashboard"].on_page_load = function (wrapper) {
	frappe.require(
		[
			"assets/apex_dashboard/js/dashboard_common.js",
			"assets/apex_dashboard/css/dashboard_common.css",
		],
		() => init_liabilities_page(wrapper)
	);
};

function init_liabilities_page(wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Liabilities Dashboard"),
		single_column: true,
	});

	const state = {
		filters: {
			company: frappe.defaults.get_user_default("company"),
			posting_date: frappe.datetime.get_today(),
		},
		data: null,
	};

	const layout = build_liabilities_layout(page);
	setup_liabilities_filters(page, state, refresh);

	function refresh() {
		page.set_indicator(__("جارِ التحميل..."), "blue");
		frappe
			.call("apex_dashboard.apex_dashboard.page.liabilities_dashboard.liabilities_dashboard.get_dashboard_data", {
				company: state.filters.company,
				posting_date: state.filters.posting_date,
			})
			.then((response) => {
				const payload = response.message || {};
				if (!payload.success) {
					throw new Error(payload.error || __("فشل تحميل بيانات الالتزامات."));
				}
				state.data = payload.data || {};
				state.filters = Object.assign({}, state.filters, payload.filters || {});
				render_liabilities_dashboard(layout, state.data);
				page.set_indicator(__("جاهز"), "green");
			})
			.catch((error) => {
				console.error(error);
				page.set_indicator(__("خطأ"), "red");
				frappe.msgprint({
					title: __("خطأ"),
					message: error.message || __("تعذر تحميل لوحة الالتزامات."),
					indicator: "red",
				});
			});
	}

	refresh();
}

function build_liabilities_layout(page) {
	const $container = $(page.body);
	$container.empty();

	const $wrapper = $(`
		<div class="gt-dashboard-wrapper">
			<section class="gt-dashboard-section">
				<div class="gt-dashboard-cards" data-role="kpis"></div>
			</section>
			<section class="gt-dashboard-section gt-dashboard-section--split">
				<div class="gt-dashboard-alerts" data-role="alerts"></div>
				<div class="gt-dashboard-section" data-role="summary"></div>
			</section>
			<section class="gt-dashboard-section" data-role="sections"></section>
		</div>
	`);

	$container.append($wrapper);

	return {
		kpis: $wrapper.find('[data-role="kpis"]')[0],
		alerts: $wrapper.find('[data-role="alerts"]')[0],
		sections: $wrapper.find('[data-role="sections"]')[0],
	};
}

function setup_liabilities_filters(page, state, refresh) {
	const company_field = page.add_field({
		fieldtype: "Link",
		fieldname: "company",
		label: __("شركة"),
		options: "Company",
		default: state.filters.company,
		change() {
			state.filters.company = company_field.get_value();
			refresh();
		},
	});

	const date_field = page.add_field({
		fieldtype: "Date",
		fieldname: "posting_date",
		label: __("تاريخ"),
		default: state.filters.posting_date,
		change() {
			state.filters.posting_date = date_field.get_value();
			refresh();
		},
	});

	page.add_action_icon("refresh", () => refresh());
}

function render_liabilities_dashboard(layout, data) {
	render_liabilities_kpis(layout.kpis, data.kpis || []);
	render_liability_alerts(layout.alerts, data.alerts || []);
	render_liability_sections(layout.sections, data.sections || []);
}

function render_liabilities_kpis(container, kpis) {
	if (!container) return;
	const cards = kpis.map((kpi) => {
		const totals = kpi.totals || {};
		const baseValue = totals.base || 0;
		const byCurrency = totals.by_currency || {};
		const subtitle = Object.keys(byCurrency)
			.map((currency) => DashboardCommon.formatCurrency(byCurrency[currency], currency))
			.join(" • ");

		return {
			title: kpi.label || "",
			value: DashboardCommon.formatCurrency(
				baseValue,
				frappe.defaults.get_default("currency") || frappe.sys_defaults.currency || "EGP"
			),
			subtitle,
			indicator: kpi.indicator || (baseValue > 0 ? "danger" : "info"),
		};
	});

	DashboardCommon.renderCards(container, cards);
}

function render_liability_alerts(container, alerts) {
	if (!container) return;
	container.innerHTML = "";

	if (!alerts.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد تنبيهات")}</div>`;
		return;
	}

	alerts.forEach((alert) => {
		const item = document.createElement("div");
		item.className = `gt-dashboard-alert level-${alert.level || "info"}`;
		item.innerHTML = `
			<span class="gt-dashboard-alert__badge">${(alert.level || "").toUpperCase()}</span>
			<span class="gt-dashboard-alert__message">${frappe.utils.escape_html(alert.message || "")}</span>
		`;
		container.appendChild(item);
	});
}

function render_liability_sections(container, sections) {
	if (!container) return;
	container.innerHTML = "";

	if (!sections.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات الالتزامات")}</div>`;
		return;
	}

	sections.forEach((section) => {
		const card = document.createElement("div");
		card.className = "gt-dashboard-subcard";

		const header = document.createElement("div");
		header.className = "gt-dashboard-section__header";
		const title = document.createElement("h4");
		title.className = "gt-dashboard-section__title";
		title.innerText = section.title || "";
		header.appendChild(title);
		header.appendChild(DashboardCommon.buildTotalsRow(section.totals || {}));
		card.appendChild(header);

		const groupsWrapper = document.createElement("div");
		groupsWrapper.className = "gt-dashboard-groups";

		(section.groups || []).forEach((group) => {
			const groupCard = document.createElement("div");
			groupCard.className = "gt-dashboard-subcard";

			const groupHeader = document.createElement("div");
			groupHeader.className = "gt-dashboard-subcard__header";
			const groupTitle = document.createElement("h5");
			groupTitle.innerText = group.label || "";
			groupHeader.appendChild(groupTitle);
			groupHeader.appendChild(DashboardCommon.buildTotalsRow(group.totals || {}));
			groupCard.appendChild(groupHeader);

			const tableContainer = document.createElement("div");
			groupCard.appendChild(tableContainer);
			DashboardCommon.renderGroupTable(tableContainer, group.balances || []);

			groupsWrapper.appendChild(groupCard);
		});

		card.appendChild(groupsWrapper);
		container.appendChild(card);
	});
}

